In [2]:
import pandas as pd

events_df = pd.read_csv('gd3/events.csv', sep=',')
purchase_df = pd.read_csv('gd3/purchase.csv', sep=',')

In [3]:
events_df.head(10)

id       event_type selected_level           start_time  tutorial_id  \
0  28903     registration            NaN  2016-05-11T23:40:55          NaN   
1  28904     registration            NaN  2016-05-11T23:49:58          NaN   
2  28905     registration            NaN  2016-05-12T00:53:07          NaN   
3  28906   tutorial_start            NaN  2016-05-12T01:32:20      17562.0   
4  28907  tutorial_finish            NaN  2016-05-12T01:34:53      17562.0   
5  28908   tutorial_start            NaN  2016-05-12T02:11:29      17563.0   
6  28909  tutorial_finish            NaN  2016-05-12T02:14:14      17563.0   
7  28910     level_choice           hard  2016-05-12T05:30:25          NaN   
8  28911      pack_choice            NaN  2016-05-12T05:38:24          NaN   
9  28912   tutorial_start            NaN  2016-05-12T07:19:01      17564.0   

   user_id  
0    12583  
1    12584  
2    12585  
3    12585  
4    12585  
5    12584  
6    12584  
7    12584  
8    12584  
9    12583

In [4]:
events_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252334 entries, 0 to 252333
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id              252334 non-null  int64  
 1   event_type      252334 non-null  object 
 2   selected_level  31086 non-null   object 
 3   start_time      252334 non-null  object 
 4   tutorial_id     125103 non-null  float64
 5   user_id         252334 non-null  int64  
dtypes: float64(1), int64(2), object(3)
memory usage: 11.6+ MB


Отобрали данные только для пользователей, зарегистрировавшихся в 2018 году.
Перевели столбец с датой в формат даты

In [5]:
cond = (events_df.start_time>='2018-01-01') & (events_df.start_time<'2019-01-01') & (events_df.event_type=='registration')
registered = events_df[cond]['user_id'].to_list() # список пользователей, зарег. в 2018
events = events_df[events_df.user_id.isin(registered)]
events.start_time = pd.to_datetime(events.start_time, format='%Y-%m-%dT%H:%M:%S')
events.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 66959 entries, 51405 to 118364
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   id              66959 non-null  int64         
 1   event_type      66959 non-null  object        
 2   selected_level  8342 non-null   object        
 3   start_time      66959 non-null  datetime64[ns]
 4   tutorial_id     32954 non-null  float64       
 5   user_id         66959 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 3.6+ MB


C:\Users\ASPIRE\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\core\generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [6]:
events.describe(include='all')

C:\Users\ASPIRE\AppData\Local\Temp/ipykernel_2480/2498275664.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  events.describe(include='all')


id    event_type selected_level           start_time  \
count    66959.000000         66959           8342                66959   
unique            NaN             5              3                66809   
top               NaN  registration         medium  2018-03-13 09:37:43   
freq              NaN         19926           4645                    9   
first             NaN           NaN            NaN  2018-01-01 03:48:40   
last              NaN           NaN            NaN  2019-01-01 05:50:36   
mean    113787.000045           NaN            NaN                  NaN   
std      19329.542752           NaN            NaN                  NaN   
min      80308.000000           NaN            NaN                  NaN   
25%      97047.500000           NaN            NaN                  NaN   
50%     113787.000000           NaN            NaN                  NaN   
75%     130526.500000           NaN            NaN                  NaN   
max     147267.000000           NaN            NaN                  NaN   

         tutorial_id       user_id  
count   32954.000000  66959.000000  
unique           NaN           NaN  
top              NaN           NaN  
freq             NaN           NaN  
first            NaN           NaN  
last             NaN           NaN  
mean    40532.934393  37781.543362  
std      5213.486632   5751.497904  
min     31505.000000  27832.000000  
25%     36008.250000  32849.000000  
50%     40529.500000  37719.000000  
75%     45057.750000  42733.000000  
max     49554.000000  47757.000000

какие уникальные события есть в колонках event_type и selected_level

In [7]:
events['event_type'].unique()

array(['registration', 'tutorial_start', 'tutorial_finish',
       'level_choice', 'pack_choice'], dtype=object)

In [8]:
events['selected_level'].unique()

array([nan, 'medium', 'hard', 'easy'], dtype=object)

какое количество пользователей совершали события:

In [9]:
events['user_id'].nunique()

19926

Читаем данные из purchase_df

In [10]:
purchase_df.head(10)

id  user_id       event_datetime  amount
0  15674    12584  2016-05-12T10:34:16     100
1  15675    12985  2016-05-13T08:25:56      50
2  15676    12828  2016-05-13T16:33:46      50
3  15677    12598  2016-05-14T01:09:37     150
4  15678    13037  2016-05-14T01:24:46     100
5  15679    12989  2016-05-14T02:19:20     200
6  15680    12684  2016-05-14T03:10:14     100
7  15681    12656  2016-05-14T03:42:48     100
8  15682    12624  2016-05-14T06:00:45      50
9  15683    13136  2016-05-14T07:31:03      50

In [11]:
purchase_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5956 entries, 0 to 5955
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              5956 non-null   int64 
 1   user_id         5956 non-null   int64 
 2   event_datetime  5956 non-null   object
 3   amount          5956 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 186.2+ KB


Отобрали данные только для пользователей, зарегистрировавшихся в 2018 году.
Перевели столбец с датой в формат даты

In [12]:
cond_2 = (purchase_df.event_datetime>='2018-01-01') & (purchase_df.event_datetime<'2019-01-01') & (events_df.event_type=='registration')
purchase = purchase_df[purchase_df.user_id.isin(registered)]
purchase.event_datetime = pd.to_datetime(purchase.event_datetime, format='%Y-%m-%dT%H:%M:%S')
purchase.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 1600 entries, 1171 to 2778
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   id              1600 non-null   int64         
 1   user_id         1600 non-null   int64         
 2   event_datetime  1600 non-null   datetime64[ns]
 3   amount          1600 non-null   int64         
dtypes: datetime64[ns](1), int64(3)
memory usage: 62.5 KB


C:\Users\ASPIRE\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\core\generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [13]:
purchase.describe()

id       user_id       amount
count   1600.000000   1600.000000  1600.000000
mean   17645.505625  37752.766250   110.734375
std      462.038637   5822.621784    54.696628
min    16845.000000  27845.000000    25.000000
25%    17245.750000  32815.750000    50.000000
50%    17645.500000  37633.500000   100.000000
75%    18045.250000  43023.000000   150.000000
max    18452.000000  47742.000000   300.000000

АНАЛИЗ ДАННЫХ. ШАГ 1

АНАЛИЗ ПОЛЬЗОВАТЕЛЬСКИХ СОБЫТИЙ

сколько пользователей совершают событие registration

In [14]:
events[events['event_type'] == 'registration']['user_id'].nunique()

19926

всего пользователей

In [15]:
events['user_id'].nunique()

19926

Как мы видим, все пользователи совершают событие registration. Это обусловлено тем, что данный этап является обязательным для выполнения и без него пользователь не сможет перейти дальше.

Посмотрим на срез данных по событию tutorial_start.

In [16]:
events[events['event_type'] == 'tutorial_start'].head(10)

id      event_type selected_level          start_time  tutorial_id  \
51412  80315  tutorial_start            NaN 2018-01-01 14:54:40      31505.0   
51413  80316  tutorial_start            NaN 2018-01-01 15:00:51      31506.0   
51415  80318  tutorial_start            NaN 2018-01-01 15:40:43      31507.0   
51417  80320  tutorial_start            NaN 2018-01-01 17:47:40      31508.0   
51420  80323  tutorial_start            NaN 2018-01-01 19:11:36      31509.0   
51422  80325  tutorial_start            NaN 2018-01-01 19:46:11      31510.0   
51434  80337  tutorial_start            NaN 2018-01-02 02:07:07      31511.0   
51435  80338  tutorial_start            NaN 2018-01-02 03:03:44      31512.0   
51437  80340  tutorial_start            NaN 2018-01-02 04:55:11      31513.0   
51446  80349  tutorial_start            NaN 2018-01-02 07:08:00      31514.0   

       user_id  
51412    27836  
51413    27835  
51415    27836  
51417    27833  
51420    27839  
51422    27834  
51434    27840  
51435    27845  
51437    27842  
51446    27845

Посмотрим на количество пользователей, которые совершают событие tutorial_start:

In [17]:
events[events['event_type'] == 'tutorial_start']['user_id'].nunique()

11858

Как мы видим, число пользователей, которые перешли к выполнению обучения, меньше, чем число пользователей, прошедших регистрацию.
19926 прошли регистрацию, 11858 переходят к обучению

Определим процент пользователей, которые перешли к выполнению обучения, и запишем его в переменную percent_tutorial_start_users.

In [18]:
registered_users_count = events[events["event_type"] == "registration"][
    "user_id"
].nunique()
tutorial_start_users_count = events[events["event_type"] == "tutorial_start"][
    "user_id"
].nunique()
percent_tutorial_start_users = tutorial_start_users_count / registered_users_count
print(
    "Процент пользователей, начавших обучение (от общего числа зарегистрировавшихся): {:.2%}".format(
        percent_tutorial_start_users
    )
)

Процент пользователей, начавших обучение (от общего числа зарегистрировавшихся): 59.51%


Теперь посмотрим, какое количество пользователей проходит обучение до конца событие tutorial_finish.

In [19]:
events[events['event_type'] == 'tutorial_finish']['user_id'].nunique()

10250

Рассчитаем процент пользователей, завершивших обучение, среди пользователей, которые начали обучение. Это будет показатель tutorial_completion_rate (коэффициент «завершаемости» обучения).

In [20]:
tutorial_finish_users_count = events[events["event_type"] == "tutorial_finish"][
    "user_id"
].nunique()
tutorial_completion_rate = tutorial_finish_users_count / tutorial_start_users_count
print(
    "Процент пользователей, завершивших обучение: {:.2%}".format(
        tutorial_completion_rate
    )
)

Процент пользователей, завершивших обучение: 86.44%


Следующий этап — это выбор уровня сложности level_choice. Давайте посмотрим на процент тех, кто доходит до этого этапа.

In [21]:
events[events['event_type'] == 'level_choice']['user_id'].nunique()

8342

Из логики нашего приложения следует, что выбор уровня сложности можно совершить и без прохождения обучения. Поэтому рассчитаем долю пользователей percent_level_choice_users, которые выбрали уровень сложности, от общего числа зарегистрировавшихся:

In [22]:
level_choice_users_count = events[events["event_type"] == "level_choice"][
    "user_id"
].nunique()
percent_level_choice_users = level_choice_users_count / registered_users_count
print(
    "Процент пользователей, выбравших уровень сложности тренировок (от общего числа зарегистрировавшихся): {:.2%}".format(
        percent_level_choice_users
    )
)

Процент пользователей, выбравших уровень сложности тренировок (от общего числа зарегистрировавшихся): 41.86%


Определим число пользователей, которые совершили событие выбора бесплатного пакета вопросов pack_choice.

In [23]:
events[events['event_type'] == 'pack_choice']['user_id'].nunique()

5737

Итак, 5737 пользователя доходят до этапа выбора бесплатных вопросов. Оценим процент таких пользователей от числа пользователей, которые выбрали уровень сложности.

In [24]:
training_choice_users_count = events[events["event_type"] == "pack_choice"][
    "user_id"
].nunique()
percent_training_choice_users = training_choice_users_count / level_choice_users_count
print(
    "Процент пользователей, выбравших набор бесплатных вопросов (от числа пользователей, которые выбрали уровень сложности): {:.2%}".format(
        percent_training_choice_users
    )
)

Процент пользователей, выбравших набор бесплатных вопросов (от числа пользователей, которые выбрали уровень сложности): 68.77%


Следующий этап после выбора бесплатного набора вопросов — это возможная покупка платных пакетов. Оценим число пользователей, которые совершили покупку.

In [26]:
purchase['user_id'].nunique()

1600

Рассчитаем процент пользователей percent_of_paying_users, которые оплатили вопросы, от числа пользователей, которые приобрели бесплатные вопросы:

In [29]:
paying_users_count = purchase["user_id"].nunique()
percent_of_paying_users = paying_users_count / training_choice_users_count
print(
    "Процент пользователей, которые оплатили вопросы (от числа пользователей, которые воспользовались бесплатными возможностями): {:.2%}".format(
        percent_of_paying_users
    )
)

Процент пользователей, которые оплатили вопросы (от числа пользователей, которые воспользовались бесплатными возможностями): 27.89%


Таким образом, чуть больше четверти пользователей, которые выбирают бесплатные вопросы, в дальнейшем покупают их.
Давайте посмотрим, какой процент составляют покупатели от общего числа зарегистрировавшихся. Для этого создадим переменную purchase_rate.

In [31]:
purchase_rate = paying_users_count / registered_users_count
print(
    "Процент пользователей, которые оплатили вопросы (от числа зарегистрировавшихся пользователей): {:.2%}".format(
        purchase_rate
    )
)

Процент пользователей, которые оплатили вопросы (от числа зарегистрировавшихся пользователей): 8.03%


АНАЛИЗ ДАННЫХ. ШАГ 2

АНАЛИЗ УНИКАЛЬНЫХ ПОЛЬЗОВАТЕЛЬСКИХ ПУТЕЙ

Объединяем датафреймы

In [33]:
purchase['event_type'] = 'purchase' #добавили столбец с только одним значением оплаты, чтобы в объединенном датафрейме четко обозначить покупку
#переименуем столбцы с id, чтобы не запутаться
events = events.rename(columns={"id": "event_id"})
purchase = purchase.rename(columns={"id": "purchase_id"})
#объединим датафреймы
total_events_df = pd.concat([events,purchase],sort=False)
total_events_df.head(10)

C:\Users\ASPIRE\AppData\Local\Temp/ipykernel_2480/3456045107.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  purchase['event_type'] = 'purchase' #добавили столбец с только одним значением оплаты, чтобы в объединенном датафрейме четко обозначить покупку


event_id       event_type selected_level          start_time  \
51405   80308.0     registration            NaN 2018-01-01 03:48:40   
51406   80309.0     registration            NaN 2018-01-01 04:07:25   
51407   80310.0     registration            NaN 2018-01-01 08:35:10   
51408   80311.0     registration            NaN 2018-01-01 11:54:47   
51409   80312.0     registration            NaN 2018-01-01 13:28:07   
51410   80313.0     registration            NaN 2018-01-01 14:08:40   
51411   80314.0     registration            NaN 2018-01-01 14:42:58   
51412   80315.0   tutorial_start            NaN 2018-01-01 14:54:40   
51413   80316.0   tutorial_start            NaN 2018-01-01 15:00:51   
51414   80317.0  tutorial_finish            NaN 2018-01-01 15:06:15   

       tutorial_id  user_id  purchase_id event_datetime  amount  
51405          NaN    27832          NaN            NaT     NaN  
51406          NaN    27833          NaN            NaT     NaN  
51407          NaN    27834          NaN            NaT     NaN  
51408          NaN    27835          NaN            NaT     NaN  
51409          NaN    27836          NaN            NaT     NaN  
51410          NaN    27837          NaN            NaT     NaN  
51411          NaN    27838          NaN            NaT     NaN  
51412      31505.0    27836          NaN            NaT     NaN  
51413      31506.0    27835          NaN            NaT     NaN  
51414      31506.0    27835          NaN            NaT     NaN

In [34]:
total_events_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68559 entries, 51405 to 2778
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   event_id        66959 non-null  float64       
 1   event_type      68559 non-null  object        
 2   selected_level  8342 non-null   object        
 3   start_time      66959 non-null  datetime64[ns]
 4   tutorial_id     32954 non-null  float64       
 5   user_id         68559 non-null  int64         
 6   purchase_id     1600 non-null   float64       
 7   event_datetime  1600 non-null   datetime64[ns]
 8   amount          1600 non-null   float64       
dtypes: datetime64[ns](2), float64(4), int64(1), object(2)
memory usage: 5.2+ MB


Воссоздаём последовательность событий.
Теперь можно для каждого пользователя создать список, который будет содержать во временной последовательности все события, совершаемые данным пользователем.

Для этого воспользуемся методом groupby по столбцу event_type и применим агрегирующую функцию apply(list). Таким образом мы сгруппируем строки по пользователю, а затем объединим в списки содержимое столбца event_type по каждому пользователю. Запишем результат в датафрейм user_path_df.

In [35]:
user_path_df = (
    total_events_df.groupby(["user_id"])["event_type"].apply(list).reset_index()
)
user_path_df.head(10)

user_id                                         event_type
0    27832                                     [registration]
1    27833    [registration, tutorial_start, tutorial_finish]
2    27834    [registration, tutorial_start, tutorial_finish]
3    27835  [registration, tutorial_start, tutorial_finish...
4    27836  [registration, tutorial_start, tutorial_start,...
5    27837                                     [registration]
6    27838                                     [registration]
7    27839  [registration, tutorial_start, tutorial_finish...
8    27840       [registration, tutorial_start, level_choice]
9    27841    [registration, tutorial_start, tutorial_finish]

Если мы объединим данные по этой колонке и подсчитаем количество пользователей по каждому пути, то получим наиболее популярные пути пользователей.

Прежде чем мы сможем оценить популярные пути пользователей, преобразуем список событий в строку event_path. Эта операция нужна для оптимизации скорости объединения, так как иначе Pandas может делать подсчёт слишком долго.

In [36]:
user_path_df["event_path"] = user_path_df["event_type"].apply(lambda x: " > ".join(x))
user_path_df["event_path"].head()

0                                         registration
1      registration > tutorial_start > tutorial_finish
2      registration > tutorial_start > tutorial_finish
3    registration > tutorial_start > tutorial_finis...
4    registration > tutorial_start > tutorial_start...
Name: event_path, dtype: object

In [38]:
user_paths = (
    user_path_df.groupby(["event_path"])["user_id"]
    .nunique()
    .sort_values(ascending=False)
)

user_paths.head(10)

event_path
registration                                                                                                       7970
registration > tutorial_start > tutorial_finish > level_choice > pack_choice                                       2796
registration > tutorial_start > tutorial_finish                                                                    1956
registration > tutorial_start > tutorial_finish > level_choice                                                     1713
registration > tutorial_start > tutorial_finish > level_choice > pack_choice > purchase                            1083
registration > tutorial_start                                                                                       842
registration > tutorial_start > level_choice > pack_choice                                                          346
registration > tutorial_start > tutorial_finish > tutorial_start > tutorial_finish                                  323
registration > tutorial_start

Давайте посмотрим, какие ещё последовательности содержат в себе оплату.

In [39]:
user_paths[user_paths.index.str.contains('purchase')].head(10)

event_path
registration > tutorial_start > tutorial_finish > level_choice > pack_choice > purchase                                                                                                             1083
registration > tutorial_start > level_choice > pack_choice > purchase                                                                                                                                124
registration > tutorial_start > tutorial_finish > level_choice > pack_choice > tutorial_start > tutorial_finish > purchase                                                                           101
registration > tutorial_start > tutorial_finish > tutorial_start > tutorial_finish > level_choice > pack_choice > purchase                                                                            52
registration > tutorial_start > tutorial_start > tutorial_finish > level_choice > pack_choice > purchase                                                                                 

Большинство последовательностей, которые содержат в себе оплату, также содержат старт обучения. Это наводит нас на гипотезу, что вероятность оплаты зависит от того, проходил ли пользователь обучение.

АНАЛИЗ ДАННЫХ. ШАГ 3

АНАЛИЗ ВРЕМЕННЫХ ПРОМЕЖУТКОВ

Определяем время между registration и tutorial_start

1. Первым делом выделим отдельный датафрейм registration_df, который будет содержать только события с event_type = registration.

In [41]:
registration_df = total_events_df[total_events_df['event_type'] == 'registration']
registration_df['user_id'].value_counts().mean()

1.0

Оставим в датафрейме registration_df только те данные, которые нужны для наших вычислений

In [42]:
registration_df = registration_df[["user_id", "start_time"]].rename(
    columns={"start_time": "registration_time"}
)

2. Выделим отдельный датафрейм tutorial_start_df, который будет содержать только события с event_type = tutorial_start (начало обучения).

In [43]:
tutorial_start_df = total_events_df[total_events_df['event_type'] == 'tutorial_start']
tutorial_start_df['user_id'].value_counts().mean()

1.522179119581717

Создадим датафрейм tutorial_start_df_wo_duplicates, где будет присутствовать только первое обучение. Для этого отсортируем датафрейм по start_time, чтобы сначала шли более ранние события начала обучения, а затем удалим дубликаты по столбцу user_id. Таким образом, для каждого user_id останется только первое событие типа tutorial_start.

In [44]:
tutorial_start_df_wo_duplicates = tutorial_start_df.sort_values(
    "start_time"
).drop_duplicates("user_id")
tutorial_start_df_wo_duplicates['user_id'].value_counts().mean()

1.0

оставим только такие столбцы, которые пригодятся нам в дальнейшем

In [45]:
tutorial_start_df_wo_duplicates = tutorial_start_df_wo_duplicates[
    ["user_id", "tutorial_id", "start_time"]
].rename(columns={"start_time": "tutorial_start_time"})

Объединяем датафреймы с первой регистрацией и первым обучением с помощью функции merge(), объединяя данные по параметру user_id.
Нужно объединить данные с использованием параметра how='inner', это позволит в объединённом датафрейме оставить только такие идентификаторы пользователей, которые есть в обоих датафреймах.

In [46]:
merged_df = registration_df.merge(
    tutorial_start_df_wo_duplicates, on="user_id", how="inner"
)
merged_df.head()

user_id   registration_time  tutorial_id tutorial_start_time
0    27833 2018-01-01 04:07:25      31508.0 2018-01-01 17:47:40
1    27834 2018-01-01 08:35:10      31510.0 2018-01-01 19:46:11
2    27835 2018-01-01 11:54:47      31506.0 2018-01-01 15:00:51
3    27836 2018-01-01 13:28:07      31505.0 2018-01-01 14:54:40
4    27839 2018-01-01 18:24:01      31509.0 2018-01-01 19:11:36

Сделаем столбец timedelta, в котором посчитаем разницу между временем начала обучения (tutorial_start_time) и временем регистрации (registration_time):

In [47]:
merged_df["timedelta"] = (
    merged_df["tutorial_start_time"] - merged_df["registration_time"]
)
merged_df.head()

user_id   registration_time  tutorial_id tutorial_start_time  \
0    27833 2018-01-01 04:07:25      31508.0 2018-01-01 17:47:40   
1    27834 2018-01-01 08:35:10      31510.0 2018-01-01 19:46:11   
2    27835 2018-01-01 11:54:47      31506.0 2018-01-01 15:00:51   
3    27836 2018-01-01 13:28:07      31505.0 2018-01-01 14:54:40   
4    27839 2018-01-01 18:24:01      31509.0 2018-01-01 19:11:36   

        timedelta  
0 0 days 13:40:15  
1 0 days 11:11:01  
2 0 days 03:06:04  
3 0 days 01:26:33  
4 0 days 00:47:35

In [48]:
merged_df["timedelta"].mean()

Timedelta('0 days 04:38:24.019817844')

In [49]:
merged_df['timedelta'].describe()

count                        11858
mean     0 days 04:38:24.019817844
std      0 days 04:15:09.650705034
min                0 days 00:00:34
25%         0 days 01:21:37.500000
50%         0 days 03:22:08.500000
75%         0 days 06:47:42.750000
max                1 days 16:03:46
Name: timedelta, dtype: object

Измеряем время обучения

In [51]:
# сформируем датафрейм tutorial_finish_df, который содержит события окончания обучения
tutorial_finish_df = total_events_df[total_events_df["event_type"] == "tutorial_finish"]

In [53]:
# сформируем список с идентификаторами первых обучений
first_tutorial_ids = tutorial_start_df_wo_duplicates["tutorial_id"].unique()

In [54]:
# отфильтруем датафрейм tutorial_finish_df, оставив в нём только события для таких обучений, которые были первыми для пользователя
tutorial_finish_df = tutorial_finish_df[
    tutorial_finish_df["tutorial_id"].isin(first_tutorial_ids)
]
# Проверим, что для каждого пользователя содержится по одному событию завершения обучения
tutorial_finish_df['user_id'].value_counts().mean()

1.0

In [55]:
tutorial_finish_df = tutorial_finish_df[["user_id", "start_time"]].rename(
    columns={"start_time": "tutorial_finish_time"}
)
merged_df_2 = tutorial_start_df_wo_duplicates.merge(
    tutorial_finish_df, on="user_id", how="inner"
)
merged_df_2["timedelta"] = (
    merged_df_2["tutorial_finish_time"] - merged_df_2["tutorial_start_time"]
)
merged_df_2.head()
print(merged_df_2["timedelta"].mean())
print(merged_df_2["timedelta"].describe())

0 days 00:03:53.174160732
count                         9830
mean     0 days 00:03:53.174160732
std      0 days 00:01:44.803217992
min                0 days 00:00:16
25%                0 days 00:02:28
50%                0 days 00:03:42
75%                0 days 00:05:08
max                0 days 00:10:06
Name: timedelta, dtype: object


Определяем время между registration и level_choice

In [56]:
level_choice_df = total_events_df[total_events_df["event_type"] == "level_choice"]
print(level_choice_df["user_id"].value_counts().mean())
level_choice_df = level_choice_df[["user_id", "start_time"]].rename(
    columns={"start_time": "level_choice_time"}
)
merged_df_3 = registration_df.merge(level_choice_df, on="user_id", how="inner")
merged_df_3["timedelta"] = (
    merged_df_3["level_choice_time"] - merged_df_3["registration_time"]
)
print(merged_df_3["timedelta"].mean())
print(merged_df_3["timedelta"].describe())

1.0
0 days 07:10:19.169863342
count                         8342
mean     0 days 07:10:19.169863342
std      0 days 04:33:51.164488800
min                0 days 00:08:15
25%         0 days 03:53:16.500000
50%                0 days 06:03:28
75%         0 days 09:34:58.500000
max                1 days 18:48:25
Name: timedelta, dtype: object


ПРОВЕРКА АНАЛИТИЧЕСКОЙ ГИПОТЕЗЫ

Определим, существует ли различие в частоте и средней величине оплат между тремя группами пользователей:
1. пользователи, которые прошли обучение хотя бы раз;
2. пользователи, которые начали обучение, но не прошли его ни разу;
3. пользователи, которые не начинали обучение, а сразу же перешли к выбору уровня сложности.

In [57]:
# Ищем пользователей, прошедших обучение хотя бы раз
users_with_finished_tutorial = total_events_df[
    total_events_df["event_type"] == "tutorial_finish"
]["user_id"].unique()
print(len(users_with_finished_tutorial))    

10250


In [58]:
# Ищем пользователей, которые начали, но не закончили обучение

# Проще всего сначала найти пользователей, у которых есть событие tutorial_start, 
# а после убрать из списка таких пользователей, у которых есть событие tutorial_finish.

users_with_started_tutorial = total_events_df[
    total_events_df["event_type"] == "tutorial_start"
]["user_id"].unique()
set_users_with_started_tutorial = set(users_with_started_tutorial)
set_users_not_finished_but_started_tutorial = (
    set_users_with_started_tutorial.difference(set(users_with_finished_tutorial))
)

print(len(set_users_with_started_tutorial))
print(len(set_users_not_finished_but_started_tutorial))
print(
    len(set_users_with_started_tutorial) - len(set(users_with_finished_tutorial))
    == len(set_users_not_finished_but_started_tutorial)
)

11858
1608
True


In [59]:
# Ищем пользователей, сразу выбравших уровень сложности

# У таких пользователей отсутствует событие tutorial_start. Поэтому мы можем просто взять 
# и убрать из множества всех пользователей множество set_users_with_started_tutorial:

all_users = total_events_df["user_id"].unique()
set_all_users = set(all_users)
set_users_not_started_tutorial = set_all_users.difference(
    set_users_with_started_tutorial
)
print(len(set_users_not_started_tutorial))
print(
    len(set_all_users) - len(set_users_with_started_tutorial)
    == len(set_users_not_started_tutorial)
)

8068
True


In [60]:
# проверка

len(set_users_not_finished_but_started_tutorial) + len(set_users_not_started_tutorial) + len(users_with_finished_tutorial) == len(set_all_users)

True

In [61]:
# Считаем пользователей, которые завершили обучение и совершили оплату

purchase_df_1 = purchase_df[purchase_df["user_id"].isin(users_with_finished_tutorial)]
purchase_df_1['user_id'].nunique()

1447

In [62]:
percent_of_purchase_1 = purchase_df_1["user_id"].nunique() / len(
    users_with_finished_tutorial
)
print(
    "Процент пользователей, которые оплатили тренировки (от числа пользователей, завершивших обучение): {:.2%}".format(
        percent_of_purchase_1
    )
)

Процент пользователей, которые оплатили тренировки (от числа пользователей, завершивших обучение): 14.12%


In [63]:
# Считаем средний размер платежа разных групп
purchase_df_1['amount'].mean()

110.98825155494126

In [64]:
purchase_df_2 = purchase_df[
    purchase_df["user_id"].isin(set_users_not_finished_but_started_tutorial)
]
print(purchase_df_2["user_id"].nunique())
percent_of_purchase_2 = purchase_df_2["user_id"].nunique() / len(
    set_users_not_finished_but_started_tutorial
)
print(
    "Процент пользователей, которые оплатили тренировки (от числа пользователей, начавших обучение, но не завершивших): {:.2%}".format(
        percent_of_purchase_2
    )
)

131
Процент пользователей, которые оплатили тренировки (от числа пользователей, начавших обучение, но не завершивших): 8.15%


In [65]:
purchase_df_2['amount'].mean() # средний чек

104.9618320610687

In [71]:
purchase_df_3 = purchase_df[purchase_df["user_id"].isin(set_users_not_started_tutorial)]
purchase_df_3['user_id'].nunique()
# количество пользователей, которые ни разу не проходили обучение

22

In [72]:
percent_of_purchase_3 = purchase_df_3["user_id"].nunique() / len(
    set_users_not_started_tutorial
)
print(
    "Процент пользователей, которые оплатили тренировки (от числа пользователей, которые ни разу не прошли обучение): {:.2%}".format(
        percent_of_purchase_3
    )
)

Процент пользователей, которые оплатили тренировки (от числа пользователей, которые ни разу не прошли обучение): 0.27%


In [74]:
purchase_df_3['amount'].mean() # средний чек у пользователей, ни разу не начавших обучение

128.4090909090909